Add ability to write findings data to bucket

dav3r opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR adds a script (and related resources) that allows an assessor to copy data to the assessment findings bucket.

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

We want to allow data to be copied from COOL assessment environments to the assessment findings bucket.

Related to:
* https://github.com/cisagov/findings-data-import-terraform/pull/34
* https://github.com/cisagov/cool-accounts/pull/132

This is part of the work required for https://github.com/cisagov/cool-system-internal/issues/122.

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

@jsf9k and I successfully applied this Terraform in a Staging environment (after I applied the Terraform in https://github.com/cisagov/findings-data-import-terraform/pull/34 and https://github.com/cisagov/cool-accounts/pull/132) and we confirmed the following:
* The new `copy-findings-data-to-bucket.sh` script was correctly generated by cloud-init.
* We were able to successfully copy a JSON file to the assessment findings bucket using `copy-findings-data-to-bucket.sh`.
* For assessments that do not require the ability to copy data to the findings bucket, the `findings_data_bucket_name` Terraform variable can be set to an empty string (`""`).

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [x] Coordinate with NOM to ensure they know how to update Production tfvars files.

jsf9k committed:
Rename write-aws-config.sh to write-terraformer-aws-config.sh

jsf9k committed:
Add an input variable for the findings data bucket name

jsf9k committed:
Add a cloud-init script to create the AWS config for the Kali instance

jsf9k committed:
Add a cloud-init script that creates the script to copy findings data

jsf9k committed:
Add cloud-init code to create the AWS credentials file and findings data copy script

dav3r committed:
Add variables related to the policy that allows assumption of the role that is allowed to write to the assessment findings bucket

dav3r committed:
Create policy that allows assumption of the role that allows writing to the assessment findings bucket and attach it to the Kali instance role

jsf9k committed:
Rework the way the Kali instance is allowed to assume delegated roles

The new way is consistent with what we did for the Terraformer
instance.

jsf9k committed:
Correct a comment to make it more accurate

jsf9k committed:
Allow the Kali instance role to assume the guac Parameter Store role

jsf9k committed:
Allow the Kali instance to use the STS and S3 endpoints

jsf9k committed:
Correct the S3 URI in the aws s3 cp command

dav3r committed:
Correct an irksome pre-existing typo

While many folks enjoy guacamole, something known as "gucamole" does not sound quite as tasty,

jsf9k committed:
Change the default value for the findings data S3 bucket name

Also update the variable description to mention the default value.

dav3r committed:
Add a missing word in a comment

dav3r committed:
Update Terraform resources in README

mcdonnnj commented on cloud-init/write-kali-aws-config.tpl.sh:
```diff
@@ -0,0 +1,79 @@
+set -o errexit
+set -o pipefail
+
+# Temporarily assume a role in order to retrieve the VNC user's
+# username from SSM Parameter Store.
+aws_sts_output=$(aws sts assume-role \
+  --role-arn="${vnc_read_parameter_store_role_arn}" \
+  --role-session-name=cloud-init)
+access_key_id=$(sed --quiet \
+  's/^[[:blank:]]*"AccessKeyId": "\([[:graph:]]\+\)",$/\1/p' \
+  <<< "$aws_sts_output")
+secret_access_key=$(sed --quiet \
+  's/^[[:blank:]]*"SecretAccessKey": "\([[:graph:]]\+\)",$/\1/p' \
+  <<< "$aws_sts_output")
+session_token=$(sed --quiet \
+  's/^[[:blank:]]*"SessionToken": "\([[:graph:]]\+\)",$/\1/p' \
+  <<< "$aws_sts_output")
+
+# Now retrieve the VNC user's username from SSM Parameter Store
+ssm_response=$(AWS_ACCESS_KEY_ID=$access_key_id \
+  AWS_SECRET_ACCESS_KEY=$secret_access_key \
+  AWS_SESSION_TOKEN=$session_token \
+  aws --region "${aws_region}" \
+  ssm get-parameter \
+  --name /vnc/username --with-decryption)
+vnc_username=$(sed --quiet \
+  's/^[[:blank:]]*"Value": "\([[:graph:]]\+\)",$/\1/p' \
+  <<< "$ssm_response")
```
If all we need is the VNC username can't we just take it as a template argument and use the value of https://github.com/cisagov/cool-assessment-terraform/blob/513726d23126a20b11c23dfb70439e1a6d90755e/locals.tf#L44-L48 like we do in https://github.com/cisagov/cool-assessment-terraform/blob/513726d23126a20b11c23dfb70439e1a6d90755e/terraformer_cloud_init.tf#L44? This also moves away from hard-coding the parameter name.

```suggestion
```

jsf9k replied:
Please see commit 7227f12.

mcdonnnj commented on cloud-init/write-terraformer-aws-config.tpl.sh:
```diff
@@ -4,7 +4,7 @@
 #
```
Since I'm assuming the VNC username bits above were sourced from this script does it also apply here?

jsf9k replied:
Please see commit d9b1e38.

mcdonnnj commented on kali_cloud_init.tf:
```diff
@@ -62,4 +62,61 @@ data "cloudinit_config" "kali_cloud_init_tasks" {
+        aws_region                          = var.aws_region
+        findings_data_bucket_write_role_arn = data.terraform_remote_state.sharedservices.outputs.assessment_findings_write_role.arn
+        permissions                         = "0400"
+        vnc_read_parameter_store_role_arn   = aws_iam_role.guacamole_parameterstorereadonly_role.arn
+        vnc_username_parameter_name         = var.ssm_key_vnc_username
```
This aligns with my above suggestions.

```suggestion
        vnc_username                        = data.aws_ssm_parameter.vnc_username.value
```

jsf9k replied:
Please see commit 7227f12.

mcdonnnj commented on cloud-init/write-kali-aws-config.tpl.sh:
```diff
@@ -0,0 +1,79 @@
+#   can be assumed to write to the findings data S3 bucket
+# * permissions - the octal permissions to assign the AWS
+#   configuration
+# * vnc_read_parameter_store_role_arn - the ARN of the role that
+#   grants read-only access to certain VNC-related SSM Parameter Store
+#   parameters, including the VNC username
+# * vnc_username_parameter_name - the name of the SSM Parameter Store
+#   parameter containing the VNC user's username
```
```suggestion
# * vnc_username - the VNC user's username
```

jsf9k replied:
Please see commit 7227f12.

mcdonnnj commented on cloud-init/write-copy-findings-data-to-bucket.tpl.sh:
```diff
@@ -0,0 +1,82 @@
+# * findings_data_bucket_name - the name of the findings data S3
+#   bucket
+# * permissions - the octal permissions to assign the script
+# * vnc_read_parameter_store_role_arn - the ARN of the role that
+#   grants read-only access to certain VNC-related SSM Parameter Store
+#   parameters, including the VNC username
+# * vnc_username_parameter_name - the name of the SSM Parameter Store
+#   parameter containing the VNC user's username
```
```suggestion
# * vnc_username - the VNC user's username
```

jsf9k replied:
Please see commit 7227f12.

mcdonnnj commented on cloud-init/write-copy-findings-data-to-bucket.tpl.sh:
```diff
@@ -0,0 +1,82 @@
+set -o errexit
+set -o pipefail
+
+# Temporarily assume a role in order to retrieve the VNC user's
+# username from SSM Parameter Store.
+aws_sts_output=$(aws sts assume-role \
+  --role-arn="${vnc_read_parameter_store_role_arn}" \
+  --role-session-name=cloud-init)
+access_key_id=$(sed --quiet \
+  's/^[[:blank:]]*"AccessKeyId": "\([[:graph:]]\+\)",$/\1/p' \
+  <<< "$aws_sts_output")
+secret_access_key=$(sed --quiet \
+  's/^[[:blank:]]*"SecretAccessKey": "\([[:graph:]]\+\)",$/\1/p' \
+  <<< "$aws_sts_output")
+session_token=$(sed --quiet \
+  's/^[[:blank:]]*"SessionToken": "\([[:graph:]]\+\)",$/\1/p' \
+  <<< "$aws_sts_output")
+
+# Now retrieve the VNC user's username from SSM Parameter Store
+ssm_response=$(AWS_ACCESS_KEY_ID=$access_key_id \
+  AWS_SECRET_ACCESS_KEY=$secret_access_key \
+  AWS_SESSION_TOKEN=$session_token \
+  aws --region "${aws_region}" \
+  ssm get-parameter \
+  --name /vnc/username --with-decryption)
+vnc_username=$(sed --quiet \
+  's/^[[:blank:]]*"Value": "\([[:graph:]]\+\)",$/\1/p' \
+  <<< "$ssm_response")
```
```suggestion
```

jsf9k replied:
Please see commit 7227f12.

mcdonnnj commented on kali_cloud_init.tf:
```diff
@@ -62,4 +62,61 @@ data "cloudinit_config" "kali_cloud_init_tasks" {
+        aws_region                        = var.aws_region
+        findings_data_bucket_name         = var.findings_data_bucket_name
+        permissions                       = "0500"
+        vnc_read_parameter_store_role_arn = aws_iam_role.guacamole_parameterstorereadonly_role.arn
+        vnc_username_parameter_name       = var.ssm_key_vnc_username
```
```suggestion
        vnc_username                      = data.aws_ssm_parameter.vnc_username.value
```

jsf9k replied:
Please see commit 7227f12.

mcdonnnj commented on kali_iam.tf:
```diff
@@ -40,3 +48,23 @@ resource "aws_iam_role_policy_attachment" "efs_mount_policy_attachment_kali" {
+    ]
+    effect = "Allow"
+    resources = [
+      aws_iam_role.guacamole_parameterstorereadonly_role.arn,
```
If the VNC username is no longer pulled from SSM I don't think we need to give this permission anymore.

```suggestion
```

jsf9k replied:
Please see commit 7227f12.

mcdonnnj commented on terraformer_cloud_init.tf:
```diff
@@ -92,18 +92,18 @@ data "cloudinit_config" "terraformer_cloud_init_tasks" {
         organization_read_role_arn                    = data.terraform_remote_state.master.outputs.organizationsreadonly_role.arn
         terraformer_role_arn                          = aws_iam_role.terraformer_role.arn
-        vnc_read_parameter_store_role_arn             = aws_iam_role.gucamole_parameterstorereadonly_role.arn
+        vnc_read_parameter_store_role_arn             = aws_iam_role.guacamole_parameterstorereadonly_role.arn
```
I believe my suggestions to rework the other templates applies here. I'm not sure if you want to break that into another PR if this is true but I wanted to at least mention it since the file's being touched.

jsf9k replied:
Please see commit d9b1e38.

mcdonnnj commented on terraformer_iam.tf:
```diff
@@ -67,7 +67,7 @@ data "aws_iam_policy_document" "terraformer_assume_delegated_role_policy_doc" {
       module.read_terraform_state.role.arn,
       data.terraform_remote_state.master.outputs.organizationsreadonly_role.arn,
-      aws_iam_role.gucamole_parameterstorereadonly_role.arn,
+      aws_iam_role.guacamole_parameterstorereadonly_role.arn,
```
If the templating changes work I'm not sure if permission to assume this role is still needed so just flagging.

jsf9k replied:
Please see commit d9b1e38.

mcdonnnj commented on variables.tf:
```diff
@@ -109,6 +121,12 @@ variable "email_sending_domains" {
   }
 }
 
+variable "findings_data_bucket_name" {
```
This is an ugly suggestion that I thought should at least be mentioned: If this being an empty string is an invalid resource should we use the `count` meta-argument hack to only create resources around this variable's use if this variable is not empty? Something like:


```terraform
count = (var.findings_data_bucket_name != "") ? 1 : 0
```

jsf9k replied:
In the case of an empty string, the policy that gets created is accepted as valid by Terraform and AWS IAM.  At the same time, the resource pointed to be the policy is invalid; therefore, there is zero possibility that the policy could accidentally allow access to an existing resource.  So I think this is OK as it is and there is no need to introduce the additional complication.  I'm open to being convinced otherwise, though.

jsf9k replied:
Please see commit 88dbf4e.

mcdonnnj reviewed (changes requested):
So on a first pass I had some somewhat complicated suggestions around the templating and the VNC username.

jsf9k committed:
Directly give Kali cloud-init scripts the VNC username

Previously these scripts were given a role ARN and an SSM Parameter
Store parameter name and used these to look up the username.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Directly give Terraformer cloud-init scripts the VNC username

Previously these scripts were given a role ARN and an SSM Parameter
Store parameter name and used these to look up the username.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Make the findings data cloud-init blocks dynamic

This allows the blocks to only be included if
var.findings_data_bucket_name is not equal to an empty string.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

mcdonnnj commented on guacamole_parameterstore_assume_role_policy_doc.tf:
```diff
@@ -14,6 +14,7 @@ data "aws_iam_policy_document" "guacamole_parameterstore_assume_role_doc" {
       identifiers = [
         aws_iam_role.guacamole_instance_role.arn,
+        aws_iam_role.kali_instance_role.arn,
         aws_iam_role.terraformer_instance_role.arn,
```
Missed this with the other suggestions but I think we can do this now.

```suggestion
```

jsf9k committed:
Stop allowing the Kali and Terraformer instance roles to assume role

These instance roles no longer need to assume the role.

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

mcdonnnj reviewed (approved):
Seems reasonable enough and LGTM ✔
